### ISSUE TITLE
Bash/Posix sh: Syntax highlighting wrong in specific `cat << EOF` case

### ISSUE THREAD

**Issue by lmerz1** [helix-editor/helix#11148]
### Summary

Given a 
```
cat <<EOF
[some text]
EOF
```
heredoc block in a Unix shell script, syntax highlighting stops working for anything after the closing `EOF` if the `[some text]`, seemingly only **in the first line of the block**, contains a single quote (`'`) **anywhere** in the line _and_ the line **starts with** a variable (`$` as line's first character).

I have tested this with different and the default themes, so it does not appear to be a theming issue.

From a quick test, my old-ish VSCodium I still keep around does not seem to exhibit this issue, so I don't think it's the language server. I might be wrong on that.

### Reproduction Steps

Save the following as `repro.sh` and open it.
```sh
usage() {
	cat <<EOF
${1}: text with quote'

more text (contents of which do not seem to matter)
	more text, indented
more text
EOF
	exit 2
}


err() {
    # some other code to illustrate incorrect syntax highlighting
    echo "hi"
}

verbose() {
    echo github
}
```
Delete either the `'`, `$`, or both in line 3 in order for the syntax highlighting to be correct again. Alternatively, add a second single quote anywhere before the closing `EOF`.

Notice how syntax highlighting also turns correct if you add another line with any non-empty content (even whitespace) above `${1}: text with quote'` – it really has to be these two characters (and with the single quote only occurring once or possibly an uneven number of times), and both occurring in the first line as described.

Screenshots:
- Buggy case:
  <details>
  <img width="738" alt="screenshot image" src="https://github.com/user-attachments/assets/172ac679-df11-4ade-b544-11dec14515d4">
  <img width="746" alt="screenshot image" src="https://github.com/user-attachments/assets/136592ab-bf65-4ff5-b087-3fd45e9129b1">
  <img width="726" alt="screenshot image" src="https://github.com/user-attachments/assets/87e13a8d-2523-4b8b-8cc0-dcee000b8f14">
   </details>
- Correct case 1: Either of the symbols not present (or at a different location)
   <details>
   <img width="727" alt="screenshot image" src="https://github.com/user-attachments/assets/686241d6-e7c3-4186-87ec-4e441c5f6510">
   <img width="712" alt="screenshot image" src="https://github.com/user-attachments/assets/ad3e4585-c3e3-4424-9a65-65324780f099">
   </details>
- Correct case 2: Added closing quote
   <details>
   <img width="709" alt="screenshot image" src="https://github.com/user-attachments/assets/013da2fe-e889-4af9-aa6d-816f6384bef9">
   </details>
- Correct case 3: Not in the first line
   <details>
   <img width="725" alt="screenshot image" src="https://github.com/user-attachments/assets/992136c6-e134-4eec-8dc1-5f8077b64447">
   </details>

Edit: added more screenshots

### Platform

macOS

### Terminal Emulator

n/a (kitty and Hyper and Terminal.app)

### Installation Method

brew

### Helix Version

helix 24.3 (2cadec0b)

**Comment by lmerz1** [comment]
Perhaps also interestingly, when adding quotes to the `EOF` symbol, which means "interpret the following block literally and do not substitute any values", the now-not-a-variable-anymore `${1}` doesn't trigger the incorrect highlighting.


<details><summary>Screenshot</summary>
<img width="731" alt="screenshot image" src="https://github.com/user-attachments/assets/6eff3d80-b32c-47f3-b3d3-046784cadf96">
</details>

**Comment by lmerz1** [comment]
Perhaps also interestingly, when adding quotes to the `EOF` symbol, which means "interpret the following block literally and do not substitute any values", the now-not-a-variable-anymore `${1}` doesn't trigger the incorrect highlighting.


<details><summary>Screenshot</summary>
<img width="731" alt="screenshot image" src="https://github.com/user-attachments/assets/6eff3d80-b32c-47f3-b3d3-046784cadf96">
</details>

**Comment by krish-r** [comment]
I believe heredoc highlighting for bash was recently fixed in #11118

**Comment by lmerz1** [comment]
Thanks. You would of course be correct. I didn't recall the name of the construct which is why searching for this issue didn't work for me. :-)
